### PR TITLE
[BREAKING][parallel] feat: support extra parallel

### PR DESCRIPTION
**Pull request by JorgenWan**
### What does this PR do?

Support extra parallel.

This parallel extends expert parallel to accommodate an arbitrary number of similar parallel patterns (e.g., embed parallel) with independent parallel size (e.g., embed parallel size).

Support FSDP1+extra_parallel and FSDP2+extra_parallel.

Support DCP checkpoint manager.

Thank [@ZZWHU](https://github.com/ZZWHU) for part of embed parallel + FSDP1.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `misc`, `ci`, `config`, `docs`, `data`, `dist`, `omni`, `logging`, `model`, `optim`, `ckpt`, `release`, `task`, `perf`, `ops`, `parallel`
  - If this PR involves multiple modules, separate them with `,` like `[ci, data, model]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][parallel, model] feat: dynamic batching`

### Test

FSDP2+extra_parallel: `tests/utils/test_extra_parallel_clip_grad_norm.py`
DCP checkpoint manager: `tests/checkpoints/test_trainer_saveload.py`

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

In the train_xxx.sh, add 
```shell
--train.accelerator.extra_parallel_sizes size1 size2
--train.accelerator.extra_parallel_placement_innermost bool1 bool2
--train.accelerator.extra_parallel_names name1 name2
```

In the modeling_xxx.py, add
```python
_no_split_modules = ["module1", "module2"]
```

In the parallel_plan.py, add 
```python
name1_plan = {
    ...
}
name2_plan = {
    ...
}
parallel_plan = ParallelPlan(
    extra_parallel_plan={
        "name1": name1_plan,
        "name2": name2_plan,
    }
)
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Comment by Luosuu** [comment]
let's discuss this at sometime. We need more careful design for maintainability I think.

File: docs/usage/basic_modules.md [hunk 1]
```diff
@@ -82,12 +82,14 @@ init_parallel_state(
     dp_replicate_size=args.train.data_parallel_replicate_size, # data parallel replicate size
```
Comment (piyifan123) [inline docs/usage/basic_modules.md]: can you add a doc to walk through the extra parallel design and key parts? it would be nice to have that for future maintenance.

File: veomni/distributed/fsdp/clip_grad_norm.py [hunk 2]
```diff
@@ -43,9 +48,10 @@ def clip_grad_norm_(fsdp_model: FSDP, max_norm, norm_type=2.0) -> torch.Tensor:
                 grads_for_clip.append(param.grad)
             # ep param
             if isinstance(spec_info.placement, Shard):
-                if ep_fsdp_process_group is None:
-                    ep_fsdp_process_group = handle.process_group
-                ep_fsdp_sharded_params_for_gnorm.setdefault(param, None)
+                if spec_info.para_mesh:
```
Comment (Luosuu) [inline veomni/distributed/fsdp/clip_grad_norm.py:51]: add inline comments to explain logic here

File: tasks/dit/train_flux.py [hunk 3]
```diff
@@ -177,11 +177,13 @@ def main():
         dp_replicate_size=args.train.data_parallel_replicate_size,
         dp_shard_size=args.train.data_parallel_shard_size,
         tp_size=args.train.tensor_parallel_size,
-        ep_size=args.train.expert_parallel_size,
         pp_size=args.train.pipeline_parallel_size,
         cp_size=args.train.context_parallel_size,
         ulysses_size=args.train.ulysses_parallel_size,
         dp_mode=args.train.data_parallel_mode,
+        extra_parallel_sizes=args.train.extra_parallel_sizes,
```
Comment (Luosuu) [inline tasks/dit/train_flux.py]: we also need to change names in job yaml?

Comment (JorgenWan) [reply tasks/dit/train_flux.py]: No, ep_size and related arguments are retained. Just view extra_parallelas an extra parallel function in the arguments.

**Comment by Luosuu** [comment]
I will take a look again later

File: veomni/distributed/torch_parallelize.py [hunk 4]
```diff
@@ -91,32 +91,53 @@ def parallelize_model_fsdp1(
...
-    Applies EP (when enabled) + FSDP1 parallel strategy to the model.
+    Applies ExtraParallel (when enabled) + FSDP1 parallel strategy to the model.
     """
-    assert not (enable_full_shard and enable_shard_grad_op), (
-        "You must explicitly specify enable_full_shard as False if enable_shard_grad_op is set to True"
-    )
     parallel_state = get_parallel_state()
 
-    if parallel_state.ep_enabled:
+    if parallel_state.any_extra_parallel_enabled:
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:102]: it seems we still need a lot of efforts to support fsdp1. maybe we can drop fsdp1 for any extra parallel at all? (we will drop fsdp1 soon anyway)

File: veomni/distributed/torch_parallelize.py [hunk 5]
```diff
@@ -353,48 +418,62 @@ def _experts_shard_placement_fn(param):
...
                 if isinstance(sub_mod, mp_ignored_classes) and sub_mod is not layer_mod:
                     fully_shard(sub_mod, **fsdp_kwargs_without_mp)
                     layer_mod._fsdp_modules.append(sub_mod)
 
-        # shard everything else in the decoder layer
-        fully_shard(layer_mod, **fsdp_kwargs)
-        layer_mod._fsdp_modules.append(layer_mod)
+        # shard everything else in the module
+        # if the module is in extra_parallel, no need to shard again
+        if not isinstance(layer_mod, FSDPModule):
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:483]: can you explain a bit more here why we need a new guard. The layer here is supposed to be decoder layer typically so we will need to shard it anyway.

File: veomni/distributed/utils.py [hunk 6]
```diff
@@ -36,6 +36,17 @@ def get_module_from_path(model: nn.Module, path: str):
         return get_module_from_path(next_obj, ".".join(attrs[1:]))
 
 
+def is_parent_module_from_path(model: nn.Module, module_name: str, path: str):
```
Comment (Luosuu) [inline veomni/distributed/utils.py:39]: is this only used for fsdp1?

File: veomni/distributed/utils.py [hunk 7]
```diff
@@ -111,3 +122,22 @@ def check_fqn_match(fqn_pattern: str, fqn: str, prefix: str = None):
     match = regex.match(fqn)
 
     return match
+
+
+def sort_fqn_by_submodule_first(fqn_list: list[str]) -> list[str]:
```
Comment (Luosuu) [inline veomni/distributed/utils.py:127]: is this only used for fsdp1?

File: veomni/distributed/torch_parallelize.py (hunk 5 above)
Comment (Luosuu) [reply veomni/distributed/torch_parallelize.py:483]: I mean please give an example here

**Comment by FoolPlayer** [comment]
LGTM, and it's better to add doc to explain the use for expert-parallel and extra-parallel in next PR.

File: veomni/distributed/utils.py (hunk 6 above)
Comment (JorgenWan) [reply veomni/distributed/utils.py:39]: yes

File: veomni/distributed/utils.py (hunk 7 above)
Comment (JorgenWan) [reply veomni/distributed/utils.py:127]: no, it is only for fsdp2

File: docs/usage/basic_modules.md (hunk 1 above)
Comment (JorgenWan) [reply docs/usage/basic_modules.md]: get

File: veomni/distributed/fsdp2/clip_grad_norm.py [hunk 8]
```diff
@@ -44,57 +45,82 @@ def clip_grad_norm(
 
 
 @torch.no_grad()
-def ep_fsdp2_clip_grad_norm(
+def extra_parallel_fsdp2_clip_grad_norm(
```
Comment (piyifan123) [inline veomni/distributed/fsdp2/clip_grad_norm.py:48]: do we have tests for this function that provides coverage for this? @Luosuu cc

File: veomni/distributed/fsdp2/clip_grad_norm.py [hunk 9]
```diff
@@ -44,57 +45,82 @@ def clip_grad_norm(
...
-        total_norm = torch.maximum(non_ep_total, ep_total)
+        total_norm = torch.maximum(non_extra_parallel_total, *extra_parallel_total.values())
     else:
-        total_norm = (non_ep_total + ep_total) ** (1.0 / float(norm_type))
+        total_norm = (non_extra_parallel_total + sum(extra_parallel_total.values())) ** (1.0 / float(norm_type))
 
     # Apply the same clip coefficient to both groups
-    torch.nn.utils.clip_grads_with_norm_(ep_params, max_norm, total_norm, foreach=foreach)
-    torch.nn.utils.clip_grads_with_norm_(non_ep_params, max_norm, total_norm, foreach=foreach)
+    for para in ps.extra_parallel_names:
```
Comment (piyifan123) [inline veomni/distributed/fsdp2/clip_grad_norm.py:121]: just to confirm the for-loop here would just loop over each group of extra parallism? e.g., we will just call this once if only fsdp2+ep?

asking because if it loops over all expert parameters that will be a lot

File: veomni/distributed/parallel_plan.py [hunk 10]
```diff
@@ -29,83 +29,126 @@
...
-
-    def apply(self, model: nn.Module, ep_fsdp_mesh: DeviceMesh):
+    def __init__(self, extra_parallel_plan: Dict[str, Dict[str, Shard]]):
+        self.extra_parallel_plan = extra_parallel_plan
+        self.extra_parallel_fsdp_no_shard_module = {
+            para_name: {".".join(list(plan.keys())[0].split(".")[:-1])}
+            for para_name, plan in self.extra_parallel_plan.items()
+        }
+
+    def apply(self, model: nn.Module, extra_parallel_fsdp_device_mesh: Dict[str, DeviceMesh]):
```
Comment (piyifan123) [inline veomni/distributed/parallel_plan.py:53]: it would be nice to have AI to backfill docstring for functions here. it's very hard to understand those logic without detailed docstring on what each function does...

Comment (piyifan123) [reply veomni/distributed/parallel_plan.py:53]: > it would be nice to have AI to backfill docstring for functions here. it's very hard to understand those logic without detailed docstring on what each function does...

same for functions in this file as well. 

(OK to do in another MR)
